File: restring/src/main/java/com/ice/restring/Restring.java
package com.ice.restring;

import android.content.Context;
import android.content.ContextWrapper;

import java.util.List;
import java.util.Map;

/**
 * Entry point for Restring. it will be used for initializing Restring components, setting new strings,
 * wrapping activity context.
 */
public abstract class Restring {

    private static boolean isInitialized = false;
    private static StringRepository stringRepository;
    private static ViewTransformerManager viewTransformerManager;

    /**
     * Initialize Restring with default configuration.
     *
     * @param context of the application.
     */
    public static void init(Context context) {
        init(context, RestringConfig.getDefault());
    }

    /**
     * Initialize Restring with the specified configuration.
     *
     * @param context of the application.
     * @param config  of the Restring.
     */
    public static void init(Context context, RestringConfig config) {
        if (isInitialized) {
            return;
        }

        isInitialized = true;
        initStringRepository(context, config);
        initViewTransformer();
    }

    /**
     * Wraps context of an activity to provide Restring features.
     *
     * @param base context of an activity.
     * @return the Restring wrapped context.
     */
    public static ContextWrapper wrapContext(Context base) {
        return RestringContextWrapper.wrap(base, stringRepository, viewTransformerManager);
    }

    /**
     * Set strings of a language.
     *
     * @param language   the strings are for.
     * @param newStrings the strings of the language.
     */
    public static void setStrings(String language, Map<String, String> newStrings) {
        stringRepository.setStrings(language, newStrings);
    }

    /**
     * Set a single string for a language.
     *
     * @param language the string is for.
     * @param key      the string key.
     * @param value    the string value.
     */
    public static void setString(String language, String key, String value) {
        stringRepository.setString(language, key, value);
    }

    static StringRepository getStringRepository() {
        return stringRepository;
    }

    private static void initStringRepository(Context context, RestringConfig config) {
        if (config.isPersist()) {
            stringRepository = new SharedPrefStringRepository(context);
        } else {
            stringRepository = new MemoryStringRepository();
        }

        if (config.getStringsLoader() != null) {
            new StringsLoaderTask(config.getStringsLoader(), stringRepository).run();
        }
    }

    private static void initViewTransformer() {
        viewTransformerManager = new ViewTransformerManager();
        viewTransformerManager.registerTransformer(new TextViewTransformer());
        viewTransformerManager.registerTransformer(new ToolbarTransformer());
        viewTransformerManager.registerTransformer(new SupportToolbarTransformer());
        viewTransformerManager.registerTransformer(new BottomNavigationViewTransformer());
    }

    /**
     * Loader of strings skeleton. Clients can implement this interface if they want to load strings on initialization.
     * First the list of languages will be asked, then strings of each language.
     */
    public interface StringsLoader {

        /**
         * Get supported languages.
         *
         * @return the list of languages.
         */
        List<String> getLanguages();

        /**
         * Get strings of a language as keys &amp; values.
         *
         * @param language of the strings.
         * @return the strings as (key, value).
         */
        Map<String, String> getStrings(String language);
    }
}

File: restring/src/main/java/com/b3nedikt/restring/Restring.kt
package com.b3nedikt.restring

import android.content.Context
import android.content.ContextWrapper
import com.b3nedikt.restring.repository.CachedStringRepository
import com.b3nedikt.restring.repository.MemoryStringRepository
import com.b3nedikt.restring.repository.SharedPrefStringRepository
import com.b3nedikt.restring.transformer.BottomNavigationViewViewTransformer
import com.b3nedikt.restring.transformer.SupportToolbarViewTransformer
import com.b3nedikt.restring.transformer.TextViewViewTransformer
import com.b3nedikt.restring.transformer.ToolbarViewTransformer
import java.util.*

/**
 * Entry point for Restring. it will be used for initializing Restring components, setting new strings,
 * wrapping activity context.
 */
object Restring {

    private var isInitialized = false
    private lateinit var stringRepository: StringRepository

    internal val viewTransformerManager: ViewTransformerManager by lazy {
        ViewTransformerManager().apply {
            registerTransformer(TextViewViewTransformer)
            registerTransformer(ToolbarViewTransformer)
            registerTransformer(SupportToolbarViewTransformer)
            registerTransformer(BottomNavigationViewViewTransformer)
        }
    }

    /**
     * Initialize Restring with the specified configuration.
     *
     * @param context of the application.
     * @param config  of the Restring.
     */
    @JvmOverloads
    fun init(context: Context, config: RestringConfig = RestringConfig.default) {
        if (isInitialized) {
            return
        }

        isInitialized = true
        initStringRepository(context, config)

        config.viewTransformers.forEach { viewTransformerManager.registerTransformer(it) }
    }

    /**
     * Wraps context of an activity to provide Restring features.
     *
     * @param base context of an activity.
     * @return the Restring wrapped context.
     */
    fun wrapContext(base: Context): ContextWrapper {
        return RestringContextWrapper.wrap(base, stringRepository)
    }

    /**
     * Set strings of a language.
     *
     * @param locale the strings are for.
     * @param newStrings the strings of the language.
     */
    fun setStrings(locale: Locale, newStrings: Map<String, String>) {
        stringRepository.setStrings(locale, newStrings)
    }

    /**
     * Set a single string for a language.
     *
     * @param locale the string is for.
     * @param key      the string key.
     * @param value    the string value.
     */
    fun setString(locale: Locale, key: String, value: String) {
        stringRepository.setString(locale, key, value)
    }

    private fun initStringRepository(context: Context, config: RestringConfig) {
        stringRepository = config.stringRepository ?: defaultRepository(context)

        if (config.stringsLoader != null) {
            StringsLoaderTask(config.stringsLoader, stringRepository).run()
        }
    }

    private fun defaultRepository(context: Context) = CachedStringRepository(
            cacheRepository = MemoryStringRepository(),
            persistentRepository = SharedPrefStringRepository(context)
    )

    /**
     * Loader of strings skeleton. Clients can implement this interface if they want to load
     * strings on initialization. First the list of languages will be asked, then strings of each
     * language.
     */
    interface StringsLoader {

        /**
         * List of supported languages
         */
        val locales: List<Locale>

        /**
         * Get strings of a language as keys &amp; values.
         *
         * @param locale of the strings.
         * @return the strings as (key, value).
         */
        fun getStrings(locale: Locale): Map<String, CharSequence>
    }
}